### ISSUE TITLE
Pointer to derived class

### ISSUE THREAD

**Issue by Udayv56** [Udayv56/CPP#38]
#include<iostream>
using namespace std;
 //Polymorphism
   // one name and multiple forms
   // Compile time - function overloading, operator overloading
   // Run time -virtual function
class BaseClass
{
  public:
    int var1;
    void display(){
        cout<<var1<<endl;
    }
};

class DerivedClass : public BaseClass{
    public:
    int var2;
    void display(){
        cout<<var1<<" "<<var2<<endl;
        
    }
};

int main(){
    BaseClass* base_class_pointer;
    BaseClass obj_base;
    DerivedClass obj_derived;
    base_class_pointer = &obj_derived; //if you will call display then it will be of base class
    base_class_pointer->var1 = 34;
    base_class_pointer->display();// pointer base class but object is of derived
    // base_class_pointer->var2 = 45; // error 
    DerivedClass* derived_class_pointer;
    derived_class_pointer = &obj_derived;
    derived_class_pointer->var2 = 45; 
    derived_class_pointer->display();
    derived_class_pointer->var1 = 405; 
    derived_class_pointer->display();
    
    return 0;
}

**OUTPUT-**

34
34 45
405 45

In the provided code, the concept of a pointer to a derived class is demonstrated. Here's an explanation of the code and the concept:

1. The code defines two classes: `BaseClass` and `DerivedClass`. `DerivedClass` is derived from `BaseClass` using the public access specifier.

2. `BaseClass` has a public member variable `var1` and a member function `display()` that prints the value of `var1`.

3. `DerivedClass` adds another member variable `var2` and overrides the `display()` function to print both `var1` and `var2`.

4. In the `main()` function, a pointer of type `BaseClass`, named `base_class_pointer`, is declared.

5. An object `obj_base` of `BaseClass` and an object `obj_derived` of `DerivedClass` are also created.

6. The statement `base_class_pointer = &obj_derived;` assigns the address of `obj_derived` to the `base_class_pointer`. This is possible because `DerivedClass` is derived from `BaseClass`, so a pointer to the base class can point to an object of the derived class.

7. The `display()` function is called using the `base_class_pointer`. However, since the function is not declared as virtual in the base class, the function resolution happens at compile time. Therefore, the `display()` function of the base class is called, even though the pointer points to an object of the derived class. This is an example of static binding or early binding.

8. Accessing `var2` using `base_class_pointer` results in an error because the pointer is of type `BaseClass`, which doesn't have `var2`. The pointer can only access the members that are part of the base class.

9. Another pointer, `derived_class_pointer`, of type `DerivedClass`, is declared and assigned the address of `obj_derived`.

10. `var2` is accessed and modified using `derived_class_pointer`, as it directly points to an object of the derived class.

11. The `display()` function is called using `derived_class_pointer`, resulting in the overridden version of the function being executed. This is an example of dynamic binding or late binding, where the function resolution is determined at runtime based on the actual object type.

12. `var1` is also modified using `derived_class_pointer`, showcasing that the pointer can access the inherited members from the base class.

In summary, a pointer to a derived class can be assigned the address of an object of the derived class. However, when accessing members or calling functions through the pointer, the behavior depends on whether the member or function is part of the base class or the derived class. Static binding occurs when accessing members/functions through a pointer of the base class type, while dynamic binding occurs when accessing members/functions through a pointer of the derived class type.
